### ISSUE TITLE
replace ANDROID_ID with a per-app random value reused across same signature apps in the profile

### ISSUE THREAD

**Issue by thestinger** [GrapheneOS/os_issue_tracker#395]
This will preserve the property of it being the same across same signature apps while no longer having it be persistent after uninstalling all same signature apps.

This is a low priority because it only prevents apps from detecting that the profile is the same after reinstallation. There are a lot of other potential ways to do this. Apps without scoped storage can generate their own ID and store it in their per-app external storage directory which persists after uninstalling and users would have to remove it to get rid of it. More generally, apps can use any data storage they're granted by the user or could use an API provided by another app.

**Comment by KikMyaz** [comment]
Would definitely love to see this get implemented. 

Certain Android OEMs seemed to already have something very similar back in Android 14 but only for certain regional variants of their devices. See screenshots. I initially thought these were referring to advertizing IDs but when I loaded my own app I realized they were not. Then I chucked the frameworks from the firmware dump into jadx and found they actually changed how `android.provider.Settings.Secure.getString("android_id")` works such that if I have the "virtual ID" thingy toggled off for any certain app it's just gonna return a bunch of zero's instead of the random 16-char(?) string. If I "reset" its value in this OEM's Settings > Privacy Protection > Virtual ID page, it actually resets that 16-char string too which was interesting. Not sure if this is helpful for GrapheneOS but it definitely deterred me from putting any trust in that value for users from those regions.

![Image](https://github.com/user-attachments/assets/c2b0bc46-475d-4478-95d4-4c6e5ddbb8d5)
![Image](https://github.com/user-attachments/assets/d0649b64-6c2e-42c7-8cf9-b7c5ec216920)

**Comment by thestinger** [comment]
The standard Android ID is already a per-app-per-profile random value but it's a persistent one.

**Comment by KikMyaz** [comment]
Makes sense. All I'm saying is there seem to be a few major OEMs that changed how that value behaves in that the user seems to be able to zero out what we get from `android.provider.Settings.Secure.getString("android_id")` on an attested device. We used to treat it as an ID pretty religiously but recently realized people have been able to get around it.

**Comment by mattcheau** [comment]
@KikMyaz What's the use case for tracking your app's installs so strictly?

**Comment by donovanphone** [comment]
@mattcheau You didn't ask me but you might find of the motivation in #7268 

The basic idea is that it's not smart to let an app know that you're using it for various purposes that need to be siloed. Like if I'm a lawyer with a certain WhatsApp Business number, do I really want Meta to also have all my personal contacts via the WhatsApp (not Business) which happens to be installed in the same profile? (I'm not saying ANDROID_ID is the cause of that happening in my case but it's one obvious route.)

With Grok, to name another app where this might well be the cause, it remembered my chat history despite clearing data and cache storage in between. At least I was actually able to observe that. Just imagine if it remembered my chats on the server side but pretended to have forgotten them in the app, and then connected them to my new chats after the storage clear. In that case, any identifying information that I happened to discuss with Grok before the clear would persist silently after the clear, connecting my later discussions with my personal identity. Like when I was asking some legal questions about how to advise my clients, which could potentially identify me if you have enough data on public court filings, and then after the clear I'm asking Grok where to find weed in my city! I'm sure you can see how ANDROID_ID undermines the siloing effect that clearing data would appear to provide. As the OP explained above, getting rid of it doesn't guarantee that siloing is suddenly airtight but at least it gets rid of this obvious and probably well exploited leakage vector.

A more reliable way to silo is to never clear data or even reinstall a given app within a given profile, the downside being that you quickly end up with a mountain of profiles. And if two apps residing within the same profile and written by two different companies suddenly merge into one company, then that merged company could have two synergistic datasets about you which were previously siloed.

Even if you had that mountain of profiles, you would still have to ensure that they don't end up sharing the same IP address, especially not simultaneously.

This is just an ugly issue all around and it probably needs to be addressed progressively, maybe including new permissions which ban access to shared or persistent storage. I would like to thank the developers in advance for working toward that, whoever they are.
